### PR TITLE
[routingconnector] convert routing processor to a connector

### PR DESCRIPTION
**Pull request by mwear**
**Description:** 

This PR introduces a connector version of the routing processor. It currently supports routing based on resource attributes OTTL statements only. Context based routing will be added later in a followup PR. 

There are two issues regarding fanout consumers that are being addressed in the core repo.
* The routing connector needs to be a consumer in multiple pipelines (the routing processor can route to a single exporter).
  * Will be resolved by: https://github.com/open-telemetry/opentelemetry-collector/pull/7840
* We need the ability to create connector routers to facilitate testing. I had to temporarily copy the fanoutconsumer package into the routing connector codebase due to package visibility issues.
  * Will be resolved by: https://github.com/open-telemetry/opentelemetry-collector/pull/7673 

We can address these issues as the relevant PRs land on the core repo.

cc: @djaglowski, @jpkrohling, @kovrus 

**Link to tracking Issue:** 
#19738

**Testing:** 
Unit / manual

**Documentation:**
The readme has been ported from the routing processor

### CODE REVIEW

File: connector/routingconnector/logs.go
Comment (djaglowski) [inline connector/routingconnector/logs.go:100]: Should we be using `ottlresource.NewTransformContext`?

File: connector/routingconnector/metrics.go
Comment (djaglowski) [inline connector/routingconnector/metrics.go:104]: What's the value of using an decrementing an int here, rather than just setting a bool when we see a match?

File: connector/routingconnector/logs.go
Comment (djaglowski) [inline connector/routingconnector/logs.go:91]: Do we need to keep track of groups according to a string key? Can we use the actual consumer? If so, I think this will simplify the `group` function.

**Review by djaglowski — COMMENTED** [review]
@mwear, thanks for taking this on. I'm excited to see so much progress.

File: connector/routingconnector/logs.go
Comment (mwear) [reply connector/routingconnector/logs.go]: I made this change in c7e8a2bf74261f8731bda4d17eaf99602dbe4a37. Instead of using `ottllog.TransformContext`, `ottlspan.TransformContext`, etc we're now using `ottlresource.TransformContext` everywhere. This allowed me to remove the second type parameter from the router. It also means our statements can refer to `attributes` instead of `resource.attributes`. If you don't mind, take a look at the commit and let me know if that's what you had in mind. It simplifies the code significantly, so long as we're not losing anything. The choice to use the signal specific `TransformContext`s was inherited from the routing processor.

File: connector/routingconnector/metrics.go
Comment (mwear) [reply connector/routingconnector/metrics.go]: Changed to a bool

File: connector/routingconnector/logs.go
Comment (mwear) [reply connector/routingconnector/logs.go]: I updated to group by consumer. It allowed me to remove the spanGroup, logGroup, etc types altogether.

File: connector/routingconnector/README.md
Comment (mwear) [reply connector/routingconnector/README.md:85]: It does. This works the same way that the routingprocessor does. Both the processor and the connector set `MutatesData: false`. I think the idea is that you can have an attribute used for routing only, and ensure that it does not get exported.

File: connector/routingconnector/logs.go
Comment (mwear) [reply connector/routingconnector/logs.go]: cc: @jpkrohling @kovrus to double check this change

**Comment by mwear** [comment]
@djaglowski thanks for the initial round of feedback. I was wondering if you had any opinions on the two issues I ran into with the fanoutconsumer (mentioned in the description). I could work around the package visibility issue by introducing  mock fanout consumers for the tests, but the actual fanout consumer is subtly complex, which introduces some risk. I have worked around the second issue, needing a connector be a receiver in multiple pipelines in order to get a fanout consumer, by adding that restriction to the routing connector; the processor did not have this restriction. Is that the right thing to do, or can we add the cabability to request a fanout consumer for a connector even if it is only a receiver in a single pipeline, the advantage being, that we can still request a consumer by name.

Let me know if you have opinions on those issues. I was waiting to get some resolution on them before taking this out of draft.

File: connector/routingconnector/logs.go
Comment (djaglowski) [reply connector/routingconnector/logs.go]: Looks right to me. We'll end up needing the signal specific contexts for other conditions but for routing based on resources, I think we can and should keep it simple.

File: connector/routingconnector/README.md
Comment (djaglowski) [reply connector/routingconnector/README.md:85]: 👍 Makes sense

**Comment by djaglowski** [comment]
@mwear, regarding problem 1 noted above:

I see what you are saying. I think what it boils down to is that we need a way for tests to create a `connector.TracesRouter` (and same for metrics/logs of course).

If we move `fanoutconsumer` from `service/internal` to `internal`, then `connectortest` could provide a minimal surface area to produce a "real" router that is only sufficient for testing. Something like the following - WDYT?

```go
type tracesRouterTestOption struct {
	id component.ID
	cons consumer.Traces
}

func WithNop(id component.ID) tracesRouterTestOption {
	return tracesRouterTestOption{ id: id, cons: consumertest.NewNop() }
}

func WithTracesSink(id component.ID) tracesRouterTestOption {
	return tracesRouterTestOption{ id: id, cons: &consumertest.TracesSink{} }	
}

func NewTracesRouterSink(opts ...tracesRouterTestOption) connector.TracesRouter {
	consumers := make(map[component.ID]consumer.Traces, numRoutes)
	for _, opt := range opts {
		consumers[opt.id] = opt.cons
	}
	return fanoutconsumer.NewTracesRouter(consumers)
}
```

**Comment by djaglowski** [comment]
Regarding problem 2:

I tend to think that a router really should expect a minimum of two downstream pipelines. It seems that anything that could be done with a router that emits to only one pipeline, could (and should) be done in a simpler way.

**Comment by mwear** [comment]
@djaglowski: Thanks for the suggestions. I like your proposal for problem 1 and will work on a PR it. I'm fine with leaving problem 2 as is. It makes sense to require multiple downstream pipelines, even though it introduce a slight difference in behavior for the connector vs the processor.

**Comment by mwear** [comment]
I'll leave this in draft until there is some resolution on: https://github.com/open-telemetry/opentelemetry-collector/pull/7673. I'll also be OOO next week (back on May 22nd). I welcome any feedback in the meantime, and will address it when I return.

File: connector/routingconnector/README.md
Comment (kovrus) [inline connector/routingconnector/README.md:76]: It seems that quite some users do routing based on the context values. Maybe after this PR we can follow up with the  OTTL function that extracts values from context if it is possible.

File: connector/routingconnector/config.go
Comment (kovrus) [inline connector/routingconnector/config.go:73]:
```suggestion
			return fmt.Errorf("invalid route %s: %w", item.Pipelines, errNoPipelines)
```

**Comment by jpkrohling** [comment]
I'm way behind my github notifications, but I'm placing this to the top of my queue.

File: connector/routingconnector/README.md
Comment (mwear) [reply connector/routingconnector/README.md:76]: We can add context based routing as a follow up, but we should discuss how we'd like that to work so that we don't paint ourselves into a corner with the first pass.

**Comment by djaglowski** [comment]
This PR is waiting for @bogdandrutu's response here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21498#issuecomment-1577924777

**Comment by mwear** [comment]
> @mwear, I think we should move forward with the temporarily copied fanout consumer code. WDYT?

That works for me. Removing the copied code will be easy once the other changes land on core. Should I dust this off and bring it out of draft?

File: connector/routingconnector/config.go
Comment (jpkrohling) [inline connector/routingconnector/config.go:51]: Both parts of this error are static. Why not just return `errNoTableItems`?

File: connector/routingconnector/config.go
Comment (jpkrohling) [inline connector/routingconnector/config.go:58]: Same to this and the next error.

File: connector/routingconnector/internal/common/functions.go
Comment (jpkrohling) [inline connector/routingconnector/internal/common/functions.go:13]: Probably a matter of taste, so, ignore me if you prefer how it is today.

```suggestion
func createRouteFunction[K any](ottl.FunctionContext, ottl.Arguments) (ottl.ExprFunc[K], error) {
```

File: connector/routingconnector/internal/fanoutconsumer/logs.go
Comment (jpkrohling) [inline connector/routingconnector/internal/fanoutconsumer/logs.go:6]: Is this a copy of what we have at the fanoutconsumer in the core? If so, explaining why this was copied and what was changed would be good.

Comment (jpkrohling) [reply connector/routingconnector/internal/fanoutconsumer/logs.go:6]: I just saw the mentions to this in the PR description. I think it needs to be in the code as well. Like: 

```
TODO: once we have fanoutconsumer test helpers, use that instead of copying the code
```

File: connector/routingconnector/logs.go
Comment (jpkrohling) [inline connector/routingconnector/logs.go:76]: Is the error related to the data, or would it err only when the OTTL statement is wrong?

File: connector/routingconnector/logs.go
Comment (jpkrohling) [inline connector/routingconnector/logs.go:93]: This might become clear later on, but isn't there a case where the only code branch reached was the error one, which made the defaultConsumer to be added already?

File: connector/routingconnector/logs.go
Comment (jpkrohling) [inline connector/routingconnector/logs.go:96]: This confused me at first, as I didn't see groups being set, but then I realized that perhaps c.group is changing groups? If that's the case, wouldn't it be better to have c.group to return groups and set groups to it? It shouldn't cause new allocations and would make the operation more explicit.

File: connector/routingconnector/logs_test.go
Comment (jpkrohling) [inline connector/routingconnector/logs_test.go:294]: Shouldn't this be an assert instead? I would even argue that all requires in this part should be asserts, as this is in the verification phase of the test, but I see why the others are `require` (just not this one)

**Review by jpkrohling — APPROVED** [review]
LGTM, a couple of minor comments, nothing blocking. This is REALLY cool, thank you @mwear!

File: .github/dependabot.yml
Comment (mwear) [reply .github/dependabot.yml:1100]: This was the result of running `make gendependabot`. I can look into it more.

File: connector/routingconnector/README.md
Comment (mwear) [reply connector/routingconnector/README.md:88]: This came from the routing processor readme: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/routingprocessor#limitations.

File: connector/routingconnector/README.md
Comment (mwear) [reply connector/routingconnector/README.md]: This actually a limitation of connectors before https://github.com/open-telemetry/opentelemetry-collector/pull/7840 merged. Since that has been merged, and released, we no longer have that limitation. I've updated the PR.

File: connector/routingconnector/config.go
Comment (mwear) [reply connector/routingconnector/config.go]: I restructured this error (and two others) in this method. Let me know if that's what you had in mind.

File: connector/routingconnector/logs.go
Comment (mwear) [reply connector/routingconnector/logs.go:76]: The `Execute` method is part of the OTTL statement. Looking at [the source](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/4e082f40991a2353d2994b376fbda1a95d51a96e/pkg/ottl/parser.go#L49-L66), it will error if there is an error evaluating the condition, or the function in the statement. I _think_ this would indicate a problem with the statement.

File: connector/routingconnector/logs.go
Comment (mwear) [reply connector/routingconnector/logs.go:93]: The error conditions would come in evaluating the routes. If no routes match (or all of them error) then the logs should be sent to the default consumer.

File: connector/routingconnector/logs.go
Comment (mwear) [reply connector/routingconnector/logs.go:96]: groups is a `map[consumer.Logs]plog.Logs`. it is mutated by `c.group`. since it is mutated by `c.group` it would be slightly confusing to me to also return and assign it. if we made that change, i would be under the impression that c.group returns a new map (without delving into the details of the method), which is not the case.

Comment (jpkrohling) [reply connector/routingconnector/logs.go:96]: Makes sense, thanks!

File: connector/routingconnector/logs.go
Comment (jpkrohling) [reply connector/routingconnector/logs.go:76]: Does this mean that this will always happen for all data points? Is it a case for a panic?

**Review by jpkrohling — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
A few comments have not been marked as resolved, but I believe all concerns have been addressed. @jpkrohling, @kovrus, please double check the open comments and raise issues if you would like to see additional changes.

@mwear, thank you for all the work on this! This is a critical connector and I'm looking forward to making it available to users.

**Pull request merged**
